### PR TITLE
feat: collect Envoy proxy configs from osm monitored pods

### PR DESCRIPTION
**Pull request by SanyaKochhar**
* Adds helper functions to run and kill a background process

### CODE REVIEW

File: pkg/collector/osmlogs_collector.go [hunk 1]
```diff
@@ -123,6 +133,61 @@ func collectPodConfigs(collector *OsmLogsCollector, rootPath, meshName, namespac
...
+		var (
+			resp    *http.Response
+			retries int = 5
+		)
+		for retries > 0 {
+			resp, err = http.Get("http://localhost:15000/config_dump")
+			if err != nil {
+				retries -= 1
+				log.Printf("Error curling config dump: %+v. %d retries remaining...\n", err, retries)
+				time.Sleep(5 * time.Second)
```
Comment (shalier) [inline pkg/collector/osmlogs_collector.go]: Why do we need to sleep here?

File: pkg/collector/osmlogs_collector.go [hunk 2]
```diff
@@ -123,6 +133,61 @@ func collectPodConfigs(collector *OsmLogsCollector, rootPath, meshName, namespac
...
+			log.Printf("Failed to write to file: %+v", err)
+			continue
+		}
+		collector.AddToCollectorFiles(resourceFile)
+
+		if err = utils.KillProcess(pid); err != nil {
+			log.Printf("Failed to kill process: %+v", err)
+			continue
+		}
+
```
Comment (shalier) [inline pkg/collector/osmlogs_collector.go]: nit: delete extra line

File: pkg/collector/osmlogs_collector.go (hunk 1 above)
Comment (SanyaKochhar) [reply pkg/collector/osmlogs_collector.go]: was facing an error where the curl was occurring before the port forwarding started, which is why there's this whole sleep and retry loop

File: deployment/aks-periscope.yaml [hunk 3]
```diff
@@ -15,8 +15,8 @@ metadata:
   name: aks-periscope-role
 rules:
 - apiGroups: ["","metrics.k8s.io"]
-  resources: ["pods", "nodes"]
-  verbs: ["get", "watch", "list"]
+  resources: ["pods", "pods/portforward", "nodes"]
```
Comment (shalier) [inline deployment/aks-periscope.yaml:18]: just curious: what's this for?

File: deployment/aks-periscope.yaml [hunk 4]
```diff
@@ -78,7 +78,7 @@ spec:
         beta.kubernetes.io/os: linux
       containers:
       - name: aks-periscope
-        image: aksrepos.azurecr.io/staging/aks-periscope:v0.3
+        image: sanyakochh/aks-periscope:latest
```
Comment (shalier) [inline deployment/aks-periscope.yaml]: should revert back?

File: deployment/aks-periscope.yaml (hunk 3 above)
Comment (SanyaKochhar) [reply deployment/aks-periscope.yaml:18]: need this access to be able to run the `kubectl port-forward` command. Don't have the right permissions without it

File: pkg/collector/osmlogs_collector.go [hunk 5]
```diff
@@ -123,7 +132,61 @@ func collectPodConfigs(collector *OsmLogsCollector, rootPath, meshName, namespac
...
+				retries -= 1
+				log.Printf("Error curling config dump: %+v. %d retries remaining...\n", err, retries)
+				time.Sleep(5 * time.Second)
+			} else {
+				break
+			}
+		}
+
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
```
Comment (SanyaKochhar) [inline pkg/collector/osmlogs_collector.go]: not sure if this handles what happens if all the retries failed. Would the resp be empty? I.e., would an `if resp != nil` block around the next section deal with that?

File: pkg/collector/osmlogs_collector.go [hunk 6]
```diff
@@ -123,7 +132,61 @@ func collectPodConfigs(collector *OsmLogsCollector, rootPath, meshName, namespac
...
+			log.Printf("Failed to collect Envoy config for pod %s in OSM monitored namespace %s: %+v", podName, namespace, err)
+			continue
+		}
+
+		var (
+			resp    *http.Response
+			retries int = 5
+		)
+		for retries > 0 {
+			resp, err = http.Get("http://localhost:15000/config_dump")
```
Comment (shashankram) [inline pkg/collector/osmlogs_collector.go]: Apart from the `config_dump` path, there are a few more that are useful:
```
"clusters",
"listeners",
"ready",
"stats",
```

File: pkg/collector/osmlogs_collector.go [hunk 7]
```diff
@@ -123,7 +132,61 @@ func collectPodConfigs(collector *OsmLogsCollector, rootPath, meshName, namespac
...
+		if err != nil {
+			log.Printf("Failed to collect Envoy config for pod %s in OSM monitored namespace %s: %+v", podName, namespace, err)
+			continue
+		}
+
+		var (
+			resp    *http.Response
+			retries int = 5
+		)
+		for retries > 0 {
```
Comment (shashankram) [inline pkg/collector/osmlogs_collector.go]: Why not a simple for loop of the form:
```
for i := 1; i <= maxRetries; i++ {
   // do something
   if success {
        break
    }

    // sleep and retry 
}
```

File: pkg/collector/osmlogs_collector.go [hunk 8]
```diff
@@ -123,7 +132,61 @@ func collectPodConfigs(collector *OsmLogsCollector, rootPath, meshName, namespac
...
+		pid, err := utils.RunBackgroundCommand("kubectl", "port-forward", podName, "-n", namespace, "15000:15000")
+		if err != nil {
+			log.Printf("Failed to collect Envoy config for pod %s in OSM monitored namespace %s: %+v", podName, namespace, err)
+			continue
+		}
+
+		var (
+			resp    *http.Response
+			retries int = 5
+		)
```
Comment (shashankram) [inline pkg/collector/osmlogs_collector.go]:
```suggestion
		var resp *http.Response
		maxRetries := 5 // some comment
```

File: pkg/collector/osmlogs_collector.go [hunk 9]
```diff
@@ -123,7 +132,61 @@ func collectPodConfigs(collector *OsmLogsCollector, rootPath, meshName, namespac
...
+
+		defer resp.Body.Close()
+		body, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			log.Printf("Failed to collect Envoy config for pod %s in OSM monitored namespace %s: %+v", podName, namespace, err)
+			continue
+		}
+		// Remove secrets from response
+		re := regexp.MustCompile("(?m)[\r\n]+^.*inline_bytes.*$")
+		secretRemovedResponse := re.ReplaceAllString(string(body), "")
```
Comment (shashankram) [inline pkg/collector/osmlogs_collector.go]:
```suggestion
		secretRemovedResponse := re.ReplaceAllString(string(body), "---redacted---")
```

**Review by shashankram — COMMENTED** [review]
Looks good! Left a few comments inline.

**Review by shalier — APPROVED** [review]

**Pull request merged**